lib/platform/linux-uart.c

```diff
@@ -0,0 +1,441 @@
+		usleep(5000);
+		rtn_rty_cnt++;
+		goto wr_rtn_retry;
+	}
+
+
+	if ((crc_exp != crc_cal) && (crc_cal != crc)){
+		usleep(5000);
+		crc_rty_cnt++;
+		goto wr_crc_err;	
+	}
+}
+
+static void uart_memcpy_to_gas(struct switchtec_dev *dev, void __gas *dest, const void *src, size_t n)
+{
+	size_t cnt;
+
+	while(n){
+		cnt = n > UART_MAX_WRITE_BYTES ? UART_MAX_WRITE_BYTES : n;
+		uart_gas_write(dev, dest, src, cnt);
+		dest += cnt;
+		src += cnt;
+		n -= cnt;   	
+	}
+}
+
+#define create_gas_write(type, suffix) \
+	static void uart_gas_write ## suffix(struct switchtec_dev *dev, \
+					type val, type __gas *addr) \
+	{ \
+		uart_gas_write(dev, addr, &val, sizeof(type)); \
+	}
+
+create_gas_write(uint8_t, 8);
+create_gas_write(uint16_t, 16);
+create_gas_write(uint32_t, 32);
+create_gas_write(uint64_t, 64);
+
+static ssize_t uart_write_from_gas(struct switchtec_dev *dev, int fd,
+				  const void __gas *src, size_t n)
+{
+	ssize_t ret;
+	void *buf;
+
+	buf = malloc(n);
+
+	uart_memcpy_from_gas(dev, buf, src, n);
+	ret = write(fd, buf, n);
+
+	free(buf);
+
+	return ret;
+}
+
+static const struct switchtec_ops uart_ops = {
+	.close = uart_close,
+	.gas_map = uart_gas_map,
+
+	.cmd = gasop_cmd,
+	.get_fw_version = gasop_get_fw_version,
+	.pff_to_port = gasop_pff_to_port,
+	.port_to_pff = gasop_port_to_pff,
+	.flash_part = gasop_flash_part,
+	.event_summary = gasop_event_summary,
+	.event_ctl = gasop_event_ctl,
+	.event_wait_for = gasop_event_wait_for,
+
+	.gas_read8 = uart_gas_read8,
+	.gas_read16 = uart_gas_read16,	
+	.gas_read32 = uart_gas_read32,
+	.gas_read64 = uart_gas_read64,
+	.gas_write8 = uart_gas_write8,
+	.gas_write16 = uart_gas_write16,
+	.gas_write32 = uart_gas_write32,
+	.gas_write64 = uart_gas_write64,
+	
+	.memcpy_to_gas = uart_memcpy_to_gas,
+	.memcpy_from_gas = uart_memcpy_from_gas,
+	.write_from_gas = uart_write_from_gas,
+};
+
+static int set_uart_attribs(int fd, int speed, int parity)
+{
+	int ret;
+	struct termios uart_attribs;
+	memset(&uart_attribs, 0, sizeof(uart_attribs));
+	
+	ret = tcgetattr(fd, &uart_attribs); 
+	if (ret)
+		return -1;
+
+	cfsetospeed(&uart_attribs, speed);
+	cfsetispeed(&uart_attribs, speed);
+
+	uart_attribs.c_iflag &= ~IGNBRK;
+	uart_attribs.c_iflag &= ~(IXON | IXOFF | IXANY);
+	uart_attribs.c_lflag = 0;
+	uart_attribs.c_oflag = 0;
+	uart_attribs.c_cflag = (uart_attribs.c_cflag & ~CSIZE) | CS8;
+	uart_attribs.c_cflag |= (CLOCAL | CREAD);
+	uart_attribs.c_cflag &= ~(PARENB | PARODD);
+	uart_attribs.c_cflag |= parity;
+	uart_attribs.c_cflag &= ~CSTOPB;
+	uart_attribs.c_cflag &= ~CRTSCTS;
+	uart_attribs.c_cc[VMIN] = 0;
+	uart_attribs.c_cc[VTIME] = 5;
+
+	ret = tcsetattr(fd, TCSANOW, &uart_attribs);
+	if (ret)
+		return -1;
+
+	return 0;
+}	
+
+struct switchtec_dev *switchtec_open_uart(const char *path)
+{
+	int ret;
+	struct switchtec_uart *udev;
+	
+	udev = malloc(sizeof(*udev));
+	if (!udev)
+		return NULL;
+
+	udev->fd = open(path, O_RDWR | O_NOCTTY);
```

Comment:
Might be a good idea to take an exclusive lock on the file as well, using flock. Then multiple switchtec calls (which are relatively safe otherwise) don't really screw each other up.

Reply:
add flock (LOCK_EX) in switchtec_open_uart and flock(LOCK_UN) in uart_close